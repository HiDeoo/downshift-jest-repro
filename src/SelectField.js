import _ from 'lodash'
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Downshift from 'downshift'

/**
 * SelectField Component.
 */
export default class SelectField extends Component {
  /**
   * React Props.
   * @type {Object}
   */
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])).isRequired,
    itemToString: PropTypes.func,
    placeholder: PropTypes.string,
  }

  /**
   * React Default Props.
   * @type {Object}
   */
  static defaultProps = {
    itemToString: item => item,
    placeholder: null,
  }

  /**
   * Creates a new instance of the component.
   * @class
   * @param {Object} props - The props of the component.
   */
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false,
    };
  }

  /**
   * Triggered when the select is toggled.
   * @param  {Object} event - The associated Downshift event.
   */
  onToggleSelect = event => {
    this.setState(({ isOpen }) => ({ isOpen: !isOpen }))
  }

  /**
   * Renders the component.
   * @return {JSX} The component.
   */
  render() {
    const { isOpen } = this.state;

    const {
      items,
      itemToString,
      placeholder,
      ...otherProps
    } = this.props;

    return (
      <Downshift isOpen={isOpen} itemToString={itemToString} {...otherProps}>
        {({
          getButtonProps,
          getInputProps,
          getItemProps,
          getRootProps,
          highlightedIndex,
          inputValue,
          selectedItem,
        }) => {
          const selectedStringValue = itemToString(selectedItem, isOpen);

          return (
            <div>
              <button
                {...getButtonProps({ onClick: this.onToggleSelect })}
                ref={button => {
                  this.button = button
                }}
              >
                {(!_.isNil(selectedItem) && !_.isEmpty(selectedItem)) ? selectedStringValue : placeholder}
              </button>
              {isOpen ? (
                <div>
                  {_.map(items, (item, index) => {
                    const stringValue = itemToString(item, isOpen)

                    return (
                      <div {...getItemProps({ item, index })} key={stringValue}>
                        {stringValue}
                      </div>
                    )
                  })}
                </div>
              ) : null}
            </div>
          )
        }}
      </Downshift>
    );
  }
}
