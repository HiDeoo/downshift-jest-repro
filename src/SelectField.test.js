import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount } from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json'

import SelectField from './SelectField';

/**
 * Enzyme configuration.
 */
Enzyme.configure({ adapter: new Adapter() })

describe('render', () => {
  test('should render properly', () => {
    const component = mount(<SelectField items={['test', 'test2']} placeholder="Choose" />)

    expect(toJSON(component)).toMatchSnapshot()
  })
});
