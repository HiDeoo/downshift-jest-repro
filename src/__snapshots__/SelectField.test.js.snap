// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render properly 1`] = `
<SelectField
  itemToString={[Function]}
  items={
    Array [
      "test",
      "test2",
    ]
  }
  placeholder="Choose"
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={null}
    getA11yStatusMessage={[Function]}
    id="downshift-1"
    isOpen={false}
    itemToString={[Function]}
    onChange={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
  >
    <div>
      <button
        aria-expanded={false}
        aria-haspopup={true}
        aria-label="open menu"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        Choose
      </button>
    </div>
  </Downshift>
</SelectField>
`;
